docs/functional_architecture.md

```diff
@@ -0,0 +1,104 @@
+Unfortunately, it is not always possible / desirable to store the resource with a simple transformation.
+
+A typical situation is that some query parameters are dynamically computed by some Javascript code to include user
+ tracking identifier, current datetime information, ...
+
+When running again the same javascript code inside the ZIM, the URL will hence be slightly different because context has changed,
+but the same content needs to be retrieved.
+
+warc2zim hence relies on fuzzy rules to transform/simplify some URLs when computing the ZIM path.
+
+## URL Rewriting
+
+warc2zim transforms (rewrites) URLs found in documents (HTML, CSS, JS, ...) so that they are usable inside the ZIM.
+
+### General case
+
+One simple example is that we might have following code in an HTML document to load an image with an absolute URL:
+
+```
+  <img src="https://en.wikipedia.org/wiki/File:Kiwix_logo_v3.svg"></img>
+```
+
+The URL `https://en.wikipedia.org/wiki/File:Kiwix_logo_v3.svg` has to be transformed to a URL that it is usable
+inside the ZIM.
+
+Since the ZIM will typically be served by kiwix-serve, this has to be a relative URL (other readers follow the same kind of conventions
+on how to access resources inside a ZIM from a given web resource). This relative URL is hence dependant on the
+location of the resource currently being rewriten.
+
+The table below gives some examples of what the rewritten URL is going to be, depending on the URL of the rewritten document.
+
+|--|--|
+| HTML document URL | image URL rewritten for usage inside the ZIM |
+| `https://en.wikipedia.org/wiki/Kiwix` | `./File:Kiwix_logo_v3.svg` |
+| `https://en.wikipedia.org/wiki` | `./wiki/File:Kiwix_logo_v3.svg` |
+| `https://en.wikipedia.org/waka/Kiwix` | `../wiki/File:Kiwix_logo_v3.svg` |
+| `https://fr.wikipedia.org/wiki/Kiwix` | `../../en.wikipedia.org/wiki/File:Kiwix_logo_v3.svg` |
+
+As can be seen on the last line (but this is true for all URLs), this rewriting has to take into account the convention saying
+ at which ZIM path a given web resource will be stored.
+
+### Dynamic case
+
+The explanation above more or less assumed that the transformations can be done statically, i.e warc2zim can open every
+known document, find existing URLs and replace them with their counterpart inside the ZIM.
+
+While this is possible for HTML and CSS documents typically, it is not possible when the URL is dynamically computed. This is
+typically the case for JS documents, where in the general case the URL is not statically stored inside the JS code but computed
+on-the-fly by aggregating various strings and values.
+
+Rewriting these computations is not deemed feasible due to the huge variety of situation which might be encountered.
+
+A specific function is hence needed to rewitre URL **live in client browser**, intercept any function triggering a web request,
+transform the URL according to conventions (where we expect the resource to be located in the general case) and fuzzy rules.
```

Comment:
We should mention Wombat (ie. external) here

Reply:
I got that, but I think it's important here to specify that this task is deferred to a third party (hence the parenthesis): this relieves the reader from keeping it in mind. But I understand so do as you feel